### PR TITLE
Enforce the same shapes

### PR DESCRIPTION
**Pull request by mcollina**
Fixes https://github.com/fastify/fastify/issues/3316

Enfore the same shape for:

* [x] Request
* [x] Reply

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

File: test/same-shape.test.js [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+const { test } = require('tap')
+const fastify = require('..')
+
+test('same shape on Request', async (t) => {
+  t.plan(1)
+
+  const app = fastify()
+
+  let request
```
Comment (jsumners) [inline test/same-shape.test.js:11]: My head is too foggy right now to understand everything from the discussion, but I wondered if this closure var affects the results?

**Review by jsumners — APPROVED** [review]
Looks good to me. I trust you followed the discussion and did your own testing. And I trust your findings.

**Comment by mcollina** [comment]
> Looks good to me. I trust you followed the discussion and did your own testing. And I trust your findings.

It also includes the tests in the PR, so we do not regress.

**Review by delvedor — APPROVED** [review]
LGTM

**Review by zekth — APPROVED** [review]
lgtm

File: test/same-shape.test.js (hunk 1 above)
Comment (mcollina) [reply test/same-shape.test.js:11]: no it doesn't in this specific case.

**Comment by RafaelGSS** [comment]
@mcollina just for curiosity, you did some benchmark before merge just to see the improvements?

**Comment by mcollina** [comment]
> @mcollina just for curiosity, you did some benchmark before merge just to see the improvements?

I checked there were no regressions. Triggering a situation where this becomes a problem requires an app more complex than our benchmarks.

There is a limited perf improvement anyway.

**Comment by darkgl0w** [comment]
Hello,

This change introduced a break in `fastify-postgres` package. It is now impossible to register named instances.
The related issue: https://github.com/fastify/fastify-postgres/issues/108

**Pull request merged**